### ISSUE TITLE
Invalid request for / (InvalidTypeException)

### ISSUE THREAD

**Issue by theroch** [icewind1991/SMB#90]
After upgrade from smb 3.1.1 to latest 3.2.3 in nextcloud server it isn't possible to access smb shares.

Because there are made a lot of modification till 3.2.3 I have no clue where I have to fix something to get it to work again.

I do not know if the error is caused by nextcloud (while using old code) or SMB?

You can find the whole stack trace [here](https://github.com/nextcloud/server/issues/20622)

**Comment by theroch** [comment]
We have one Ubuntu Server 16.04 with php 7.3 and Nextcloud 18.0.4 where the smb shares are working properly.

The other server is Ubuntu Server 18.04 with php 7.4 and Nextcloud 18.0.4 where the smb shares are not working.

The problem is caused by
smb/src/Native/NativeFileInfo.php
in the isDirectory method.
The mode for a directory on the non working system is always 0x41ed (16877). The bit operation with 0x10 (for directory) in the new icewind lib is therfore always 0.
If I use the old method !(0x41ed & 0x20) all is working again correct.

Try this fix in isDirectory Line 112:
`return !(bool)($mode & 0x20);`

The mode on the working system is always 0x10 (16).
On the non working system the mode is always 0x41ed (16877)
The mode is read with this code:
smbclient_getxattr($state, $uri, 'system.dos_attr.*');

You can test the mode on your system with the following code:
```
<?php
$workgroup ='your workgroup or domain';
$user           ='your user';
$pwd           = 'your secret password';
$uri             = 'smb://your.server.local/share';
// Create new state:
$state = smbclient_state_new();
// Initialize the state with workgroup, username and password:
smbclient_state_init($state, $workgroup, $user, $pwd);

$attr = @smbclient_getxattr($state, $uri, 'system.dos_attr.*');
var_dump($attr);
```

**Comment by theroch** [comment]
[icewind1991/SMB@927e330](https://github.com/icewind1991/SMB/commit/927e330901cf7d1323937a2e77e6d542fe9cd894) seems to fix it.

How did you come up with the unix representation?
This is the best explanation for the behaviour so far.
But it's still a little weird for me:
`0x41ed` = `40755` in Oct
But my directory mode is `40777` = `0x41FF`
My SMB config is:
```
force directory mode = 0775
force create mode = 0775
```
so it shouldbe 0x41FD?

**Comment by icewind1991** [comment]
> 0x41ed = 40755 in Oct
But my directory mode is 40777 = 0x41FF

From how I understand it the "unix mode" returned isn't the unix mode of the file on the server's filesystem, the file mode is first converted to to smb permissions by the samba server and then converted back to unix mode by libsmbclient.

Since smb/dos permissions don't have the concept of user/group/world permissions it only sets the user writable bit

**Comment by icewind1991** [comment]
should be fixed with 3.2.4
